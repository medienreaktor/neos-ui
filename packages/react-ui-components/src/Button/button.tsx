import mergeClassNames from 'classnames';
import React from 'react';

import {Omit, PickDefaultProps} from '../../types';
import {makeFocusNode} from '../_lib/focusNode';

export type ButtonStyle = 'clean' | 'brand' | 'lighter' | 'transparent' | 'warn';
export type ButtonHoverStyle = 'clean' | 'brand' | 'darken' | 'warn';
export type ButtonSize = 'small' | 'regular';

interface ButtonTheme {
    readonly 'btn': string;
    readonly 'btn--clean': string;
    readonly 'btn--lighter': string;
    readonly 'btn--transparent': string;
    readonly 'btn--brand': string;
    readonly 'btn--brandActive': string;
    readonly 'btn--brandHover': string;
    readonly 'btn--cleanHover': string;
    readonly 'btn--isPressed': string;
    readonly 'btn--darkenHover': string;
}

// We omit the standard HTML button style attribute,
// so we have no collision with the Button component's style property,
// while still enjoying the intellisense and type checking for the rest of the HTML button attributes
export type HTMLButtonElementAttributesExceptStyle = Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'style'>;

// own props and (optional) HTML button attributes except 'style'
export interface ButtonProps extends HTMLButtonElementAttributesExceptStyle {
    /**
     * This prop controls the visual pressed state of the `Button`.
     */
    readonly isPressed?: boolean;

    /**
     * This prop controls the visual focused state of the `Button`.
     * When `true`, the node gets focused via the DOM API.
     */
    readonly isFocused?: boolean;

    /**
     * This prop controls the visual and interactive disabled state of the `Button`.
     * When `true`, the node gets rendered with a truthy `disabled` prop.
     */
    readonly isDisabled?: boolean;

    /**
     * This prop controls the visual active state of the `Button`.
     */
    readonly isActive?: boolean;

    /**
     * The `kind` prop defines the regular visual style of the `Button`.
     */
    readonly style?: ButtonStyle;

    /**
     * As the `style` prop, this prop controls the visual :hover style of the `Button`.
     */
    readonly hoverStyle?: ButtonHoverStyle;

    /**
     * Defines the size of the button.
     */
    readonly size: ButtonSize;

    /**
     * An optional `className` to attach to the wrapper.
     */
    readonly className?: string;

    /**
     * The contents to be rendered within the `Bar`.
     */
    readonly children: React.ReactNode;

    /**
     * An optional css theme to be injected.
     */
    readonly theme?: ButtonTheme;

    /**
     * An interal prop for testing purposes, do not set this prop manually.
     */
    readonly _refHandler?: (isFocused: boolean) => (node: any) => void;
}

type DefaultProps = PickDefaultProps<ButtonProps,
    '_refHandler' |
    'hoverStyle' |
    'isActive' |
    'isDisabled' |
    'isFocused' |
    'size' |
    'style' |
    'type'
>;

export const defaultProps: DefaultProps = {
    _refHandler: makeFocusNode,
    hoverStyle: 'brand',
    isActive: false,
    isDisabled: false,
    isFocused: false,
    size: 'regular',
    style: 'lighter',
    type: 'button',
};

class Button extends React.PureComponent<ButtonProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element {
        const {
            children,
            className,
            isPressed,
            isFocused,
            isDisabled,
            isActive,
            style,
            hoverStyle,
            size,
            theme,
            type,
            _refHandler,
            ...rest
        } = this.props;
        const effectiveStyle = isActive ? 'brand' : style;
        const effectiveHoverStyle = isActive ? 'brand' : hoverStyle;
        const finalClassName = mergeClassNames(
            theme!.btn,
            // @ts-ignore implizit any because ButtonTheme has no index signature
            theme![`btn--size-${size}`],
            // @ts-ignore implizit any because ButtonTheme has no index signature
            theme![`btn--${effectiveStyle!}`],
            // @ts-ignore implizit any because ButtonTheme has no index signature
            theme![`btn--${effectiveHoverStyle!}Hover`],
            {
                [theme!['btn--brandActive']]: isActive,
                [theme!['btn--isPressed']]: isPressed,
            },
            className,
        );

        return (
            <button {...rest} disabled={isDisabled} type={type} className={finalClassName} role="button" ref={_refHandler && _refHandler(isFocused!)}>
                {children}
            </button>
        );
    }
}

export default Button;
