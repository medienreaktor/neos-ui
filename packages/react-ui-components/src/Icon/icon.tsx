import {IconName, IconPrefix, IconProp} from '@fortawesome/fontawesome-svg-core';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import mergeClassNames from 'classnames';
import React, {PureComponent} from 'react';

import {PickDefaultProps} from '../../types';
import mapper from './mapper';

type IconPadding = 'none' | 'left' | 'right';
type IconColor = 'default' | 'warn' | 'error' | 'primaryBlue';

interface IconTheme {
    readonly icon: string;
    readonly 'icon--big': string;
    readonly 'icon--small': string;
    readonly 'icon--tiny': string;
    readonly 'icon--paddedLeft': string;
    readonly 'icon--paddedRight': string;
    readonly 'icon--spin': string;
    readonly 'icon--color-warn': string;
    readonly 'icon--color-error': string;
    readonly 'icon--color-primaryBlue': string;
}

export interface IconProps {
    /**
     * We use the react component FortAwesome provides to render icons.
     * we will pass down all props to the component via {...rest} to expose it's api
     * https://github.com/FortAwesome/react-fontawesome/blob/master/README.md
     */

    /**
     * The ID of the icon to render.
     */
    readonly icon: string;

    /**
     * The (accessibility) label for this icon
     */
    readonly label?: string;

    /**
     * Controls the padding around the icon in a standardized way.
     */
    readonly padded?: IconPadding;

    /**
     * An optional `className` to attach to the wrapper.
     */
    readonly className?: string;

    /**
     *  Adjust the color of the icon
     */
    readonly color?: IconColor;

    /**
     *  An optional css theme to be injected.
     */
    readonly theme?: IconTheme;
}

type DefaultProps = PickDefaultProps<IconProps, 'color' | 'padded'>;

export const defaultProps: DefaultProps = {
    color: 'default',
    padded: 'none'
};

class Icon extends PureComponent<IconProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element |Â null {
        const {padded, theme, label, icon, className, color, ...rest} = this.props;
        const iconClassName = icon;
        const classNames = mergeClassNames(
            theme!.icon,
            iconClassName,
            className,
            {
                [theme!['icon--paddedLeft']]: padded === 'left',
                [theme!['icon--paddedRight']]: padded === 'right',
                [theme!['icon--color-warn']]: color === 'warn',
                [theme!['icon--color-error']]: color === 'error',
                [theme!['icon--color-primaryBlue']]: color === 'primaryBlue',
            }
        );

        return <FontAwesomeIcon icon={this.getIconProp(icon) || 'question'} aria-label={label} className={classNames} {...rest} />;
    }

    private readonly getIconProp = (icon: string): IconProp => {
        const mappedIcon = mapper(icon);
        const iconArray = mappedIcon.split(' ');
        if (iconArray.length > 1) {
            const prefix = iconArray[0];
            const processedIcon = iconArray[1].startsWith('fa-') ? iconArray[1].substr(3) : iconArray[1];
            return [prefix as IconPrefix, processedIcon as IconName];
        } else {
            const prefix: IconPrefix = 'fas';
            return [prefix, mappedIcon as IconName];
        }
    }
}

export default Icon;
