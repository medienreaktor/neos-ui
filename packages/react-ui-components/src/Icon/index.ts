import {config, IconPrefix, library} from '@fortawesome/fontawesome-svg-core';
import '@fortawesome/fontawesome-svg-core/styles.css';
import {fab} from '@fortawesome/free-brands-svg-icons';
import {far} from '@fortawesome/free-regular-svg-icons';
import {fas} from '@fortawesome/free-solid-svg-icons';
import {themr} from '@friendsofreactjs/react-css-themr';

import identifiers from '../identifiers';
import Icon from './icon';

import style from './style.css';

// tslint:disable:no-object-mutation
config.autoAddCss = false;
config.familyPrefix = 'neos-fa' as IconPrefix;
config.replacementClass = 'neos-svg-inline--fa';
// tslint:enable:no-object-mutation

library.add(fab, fas, far);

export default themr(identifiers.icon, style)(Icon);
