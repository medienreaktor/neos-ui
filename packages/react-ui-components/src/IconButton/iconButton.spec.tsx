import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import Button from '../Button';
import Icon from '../Icon';
import IconButton, {defaultProps, IconButtonProps} from './iconButton';

describe('<IconButton/>', () => {
    const props: IconButtonProps = {
        ...defaultProps,
        className: 'fooClassName',
        icon: 'fooIconName',
        theme: {
            'iconButton': 'iconButtonClassName',
            'iconButton--disabled': 'disabledClassName',
        }
    };

    it('should render correctly.', () => {
        const wrapper = shallow(<IconButton {...props}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should allow the propagation of "className" with the "className" prop.', () => {
        const wrapper = shallow(<IconButton {...props}/>);

        expect(wrapper.prop('className')).toContain('fooClassName');
    });

    it('should have "disabledClassName" set in "className" prop when "disabled" props is set to "true".', () => {
        const wrapper = shallow(<IconButton {...props} disabled={true}/>);

        expect(wrapper.prop('className')).toContain('disabledClassName');
    });

    it('should render the "Button" component and propagate all props to it.', () => {
        const wrapper = shallow(<IconButton {...props}/>);
        const btn = wrapper.find(Button);

        expect(btn.hasClass('fooClassName')).toBeTruthy();
    });

    it('should render the "Icon" component and propagate the "icon" prop to it.', () => {
        const wrapper = shallow(<IconButton {...props}/>);
        const icon = wrapper.find(Icon);

        expect(icon.prop('icon')).toBe('fooIconName');
    });
});
