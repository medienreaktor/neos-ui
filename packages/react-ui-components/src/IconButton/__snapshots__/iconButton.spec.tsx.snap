// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton/> should render correctly. 1`] = `
<ThemedButton
  _refHandler={[Function]}
  className="fooClassName iconButtonClassName"
  composeTheme="deeply"
  hoverStyle="brand"
  isActive={false}
  isDisabled={false}
  isFocused={false}
  mapThemrProps={[Function]}
  size="regular"
  style="transparent"
  type="button"
>
  <ThemedIcon
    color="default"
    composeTheme="deeply"
    icon="fooIconName"
    mapThemrProps={[Function]}
    padded="none"
  />
</ThemedButton>
`;
